server/app/views/admin/programs/ProgramBlockView.java

```diff
@@ -1,98 +1,844 @@
+                .build());
+    return qb.getContainer(questionBankVisibility);
+  }
+
+  private Modal blockDescriptionModal(
+      InputTag csrfTag, BlockForm blockForm, String blockUpdateAction) {
+    String modalTitle = "Screen name and description";
+    FormTag blockDescriptionForm =
+        form(csrfTag).withMethod(HttpVerbs.POST).withAction(blockUpdateAction);
+    blockDescriptionForm
+        .withId("block-edit-form")
+        .with(
+            div(
+                    h1("The following fields will only be visible to administrators")
+                        .withClasses("text-base", "mb-2"),
+                    FieldWithLabel.input()
+                        .setId("block-name-input")
+                        .setFieldName("name")
+                        .setLabelText("Screen name")
+                        .setValue(blockForm.getName())
+                        .getInputTag(),
+                    FieldWithLabel.textArea()
+                        .setId("block-description-textarea")
+                        .setFieldName("description")
+                        .setLabelText("Screen description")
+                        .setValue(blockForm.getDescription())
+                        .getTextareaTag())
+                .withClasses("mx-4"),
+            submitButton("Save")
+                .withId("update-block-button")
+                .withClasses(
+                    "mx-4", "my-1", "inline", "opacity-100", StyleUtils.disabled("opacity-50"))
+                .isDisabled());
+    ButtonTag editScreenButton =
+        ViewUtils.makeSvgTextButton("Edit screen name and description", Icons.EDIT)
+            .withClasses(AdminStyles.SECONDARY_BUTTON_STYLES);
+    return Modal.builder("block-description-modal", blockDescriptionForm)
+        .setModalTitle(modalTitle)
+        .setTriggerButtonContent(editScreenButton)
+        .setWidth(Modal.Width.THIRD)
+        .build();
+  }
+
+  private boolean hasNoRepeatedBlocks(ProgramDefinition programDefinition, long blockId) {
+    return programDefinition.getBlockDefinitionsForEnumerator(blockId).isEmpty();
+  }
+
+  private boolean viewAllowsEditingProgram() {
+    return programDisplayType.equals(DRAFT);
+  }
+
+  @Override
+  protected String getEditButtonText() {
+    if (viewAllowsEditingProgram()) {
+      return "Edit program details";
+    } else {
+      return "Edit Program";
     }
+  }
 
-    container.with(
-        text(blockName + " is " + predicateDefinition.action().toDisplayString() + " any of:"));
-    UlTag conditionList = ul().withClasses("list-disc", "ml-4", "mb-4");
+  @Override
+  protected String getEditButtonUrl(ProgramDefinition programDefinition) {
+    if (viewAllowsEditingProgram()) {
+      return routes.AdminProgramController.edit(programDefinition.id()).url();
+    }
+    // TODO(#3162) add the route once a read only navigation option is available
+    return NOT_YET_IMPLEMENTED_ERROR_MESSAGE;
```

Comment:
**issue (blocking):** Same as above, throw an exception rather than create a silent error.

Reply:
I believe that this is good enough because the class itself throws an exception when it is created "in the wrong state". My other change is ready to replace this (pending review).
It also matches your original suggestion of returning an obviously wrong value rather than using the edit version.